File: step_04_binary_search/4.2_bs_on_answers/P009_book_allocation_problem/Solution.java
import java.util.*;

// ============================================================
// APPROACH 1: BRUTE FORCE
// Time: O(n * sum(pages))  |  Space: O(1)
// Try every candidate for max pages from max(pages) to sum(pages).
// For each, greedily count how many students are needed.
// Return the smallest valid candidate.
// ============================================================
class BruteForce {
    private static int studentsNeeded(int[] pages, int maxPages) {
        int students = 1, cur = 0;
        for (int p : pages) {
            if (cur + p > maxPages) { students++; cur = p; }
            else cur += p;
        }
        return students;
    }

    public static int solve(int[] pages, int m) {
        if (m > pages.length) return -1;
        int lo = Arrays.stream(pages).max().getAsInt();
        int hi = Arrays.stream(pages).sum();
        for (int limit = lo; limit <= hi; limit++) {
            if (studentsNeeded(pages, limit) <= m) return limit;
        }
        return -1;
    }
}

// ============================================================
// APPROACH 2: OPTIMAL - Binary Search on Answer
// Time: O(n log(sum))  |  Space: O(1)
// Binary search between max(pages) and sum(pages).
// For each mid, check if m students suffice. Find smallest valid mid.
// ============================================================
class Optimal {
    private static boolean isPossible(int[] pages, int m, int maxPages) {
        int students = 1, cur = 0;
        for (int p : pages) {
            if (p > maxPages) return false;
            if (cur + p > maxPages) { students++; cur = p; }
            else cur += p;
        }
        return students <= m;
    }

    public static int solve(int[] pages, int m) {
        if (m > pages.length) return -1;
        int lo = Arrays.stream(pages).max().getAsInt();
        int hi = Arrays.stream(pages).sum();
        int ans = hi;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            if (isPossible(pages, m, mid)) { ans = mid; hi = mid - 1; }
            else lo = mid + 1;
        }
        return ans;
    }
}

// ============================================================
// APPROACH 3: BEST - Binary search with lo<hi loop form
// Time: O(n log(sum))  |  Space: O(1)
// Same algorithm with lower-mid form for minimization problems
// ============================================================
class Best {
    private static boolean feasible(int[] pages, int m, int limit) {
        int students = 1, cur = 0;
        for (int p : pages) {
            if (p > limit) return false;
            if (cur + p > limit) { students++; cur = p; }
            else cur += p;
            if (students > m) return false;
        }
        return true;
    }

    public static int solve(int[] pages, int m) {
        if (m > pages.length) return -1;
        int lo = Arrays.stream(pages).max().getAsInt();
        int hi = Arrays.stream(pages).sum();
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2; // lower mid for minimization
            if (feasible(pages, m, mid)) hi = mid;
            else lo = mid + 1;
        }
        return lo;
    }
}

public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Book Allocation Problem ===");

        int[][] tests   = {{12, 34, 67, 90}, {10, 20, 30, 40}};
        int[]   ms      = {2, 2};
        int[]   expected = {113, 60};

        for (int t = 0; t < tests.length; t++) {
            int[] pages = tests[t];
            int m = ms[t];
            System.out.printf("Pages=%s, m=%d => Brute=%d, Optimal=%d, Best=%d (exp=%d)%n",
                Arrays.toString(pages), m,
                BruteForce.solve(pages, m),
                Optimal.solve(pages, m),
                Best.solve(pages, m),
                expected[t]);
        }
    }
}

File: step_04_binary_search/4.2_bs_on_answers/P009_book_allocation_problem/solution.py
"""Problem: Book Allocation - Minimize the maximum pages allocated to any one student"""

# ============================================================
# APPROACH 1: BRUTE FORCE
# Time: O(n * sum(pages))  |  Space: O(1)
# Try every value from max(pages) to sum(pages) as the limit.
# Return the first limit where m students suffice.
# ============================================================
def brute_force(pages: list, m: int) -> int:
    if m > len(pages):
        return -1

    def students_needed(limit):
        students, cur = 1, 0
        for p in pages:
            if cur + p > limit:
                students += 1
                cur = p
            else:
                cur += p
        return students

    for limit in range(max(pages), sum(pages) + 1):
        if students_needed(limit) <= m:
            return limit
    return -1


# ============================================================
# APPROACH 2: OPTIMAL - Binary Search on Answer
# Time: O(n log(sum))  |  Space: O(1)
# Binary search between max(pages) and sum(pages). For each mid,
# greedily check if allocation is possible with m students.
# ============================================================
def optimal(pages: list, m: int) -> int:
    if m > len(pages):
        return -1

    def is_possible(limit):
        students, cur = 1, 0
        for p in pages:
            if p > limit:
                return False
            if cur + p > limit:
                students += 1
                cur = p
            else:
                cur += p
        return students <= m

    lo, hi, ans = max(pages), sum(pages), sum(pages)
    while lo <= hi:
        mid = (lo + hi) // 2
        if is_possible(mid):
            ans = mid
            hi = mid - 1
        else:
            lo = mid + 1
    return ans


# ============================================================
# APPROACH 3: BEST - Binary search lo<hi form (minimization template)
# Time: O(n log(sum))  |  Space: O(1)
# Clean lo<hi with lower-mid — standard pattern for "minimize the max"
# ============================================================
def best(pages: list, m: int) -> int:
    if m > len(pages):
        return -1

    def feasible(limit):
        students, cur = 1, 0
        for p in pages:
            if p > limit:
                return False
            if cur + p > limit:
                students += 1
                cur = p
                if students > m:
                    return False
            else:
                cur += p
        return True

    lo, hi = max(pages), sum(pages)
    while lo < hi:
        mid = (lo + hi) // 2  # lower mid for minimization
        if feasible(mid):
            hi = mid
        else:
            lo = mid + 1
    return lo


if __name__ == "__main__":
    print("=== Book Allocation Problem ===")
    tests = [
        ([12, 34, 67, 90], 2, 113),
        ([10, 20, 30, 40], 2, 60),
        ([15, 17, 20], 2, 32),  # optimal: [15,17] and [20] => max=32
    ]
    for pages, m, expected in tests:
        b  = brute_force(pages, m)
        o  = optimal(pages, m)
        be = best(pages, m)
        print(f"Pages={pages}, m={m} => Brute={b}, Optimal={o}, Best={be} (exp={expected})")
        assert b == o == be == expected, f"Mismatch: {b} {o} {be}"
    print("All tests passed.")
